### ISSUE TITLE
The doc "deploying your own personal vpn with OpenVPN in AWS" could be improved for beginners

### ISSUE THREAD

**Issue by louisremi** [bonovoxly/playbook#3]
This page ranks third on Google for the query "openvpn docker aws", so many people will find it before finding your `playbook` and `containers` repos. Which means the « quickstart starting from zero » section is absolutely not... starting from zero :-)

I'll update this issue with the steps I had to figure out by myself before being able to use openvpn on aws:

- make sure I had generated an ssh-key named `awskey` (and not anything else)
- added the awscli package to `environment.sh` in my `containers` clone
- follow the quickstart instructions in `containers`'README (and figure out the virtualenv stuff)
- run `source ../containers/env/bin/activate` from within `playbook/ansible-playbooks`
- saving a password in `~/.ansible/vault_deploy_key`
- create the vault file using `ansible-vault create vars/dev_aws/vault.yml`
- read the vpc playbook using `ansible-playbook vpc_create.yml -e "env=dev"` instead of `ansible-playbook create_vpc.yml -e "env=dev"`
- upload a public key to the appropriate region of ec2, named "dev"
- read the openvpn playbook using `ansible-playbook openvpn.yml -e "env=dev assign_public_ip=true instance_type=t2.micro"` instead of `ansible-playbook openvpn.yml -e "env=dev public_ip=true instance_type=t2.micro"`

**Comment by louisremi** [comment]
I'm now stuck with the following error when executing `ansible-playbook openvpn.yml -e "env=dev public_ip=true instance_type=t2.micro"`:
```
TASK [aws.securitygroups : Create security groups.] ****************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: <Response><Errors><Error><Code>InvalidParameterValue</Code><Message>CIDR block /32 is malformed</Message></Error></Errors><RequestID>8f8f2151-5fc2-4a98-9b5a-82345f7c722d</RequestID></Response>
failed: [127.0.0.1] (item={u'rules': [{u'to_port': 1194, u'from_port': 1194, u'cidr_ip': u'0.0.0.0/0', u'proto': u'udp'}], u'description': u'openvpn inbound rules', u'name': u'openvpn_inbound', u'tags': {u'Environment': u'dev', u'Name': u'openvpn_inbound'}}) => {"failed": true, "item": {"description": "openvpn inbound rules", "name": "openvpn_inbound", "rules": [{"cidr_ip": "0.0.0.0/0", "from_port": 1194, "proto": "udp", "to_port": 1194}], "tags": {"Environment": "dev", "Name": "openvpn_inbound"}}, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_aKVUNz/ansible_module_ec2_group.py\", line 478, in <module>\n    main()\n  File \"/tmp/ansible_aKVUNz/ansible_module_ec2_group.py\", line 373, in main\n    group.authorize(rule['proto'], rule['from_port'], rule['to_port'], thisip, grantGroup)\n  File \"/home/louisremi/Workspace/containers/env/local/lib/python2.7/site-packages/boto/ec2/securitygroup.py\", line 203, in authorize\n    dry_run=dry_run)\n  File \"/home/louisremi/Workspace/containers/env/local/lib/python2.7/site-packages/boto/ec2/connection.py\", line 3207, in authorize_security_group\n    params, verb='POST')\n  File \"/home/louisremi/Workspace/containers/env/local/lib/python2.7/site-packages/boto/connection.py\", line 1227, in get_status\n    raise self.ResponseError(response.status, response.reason, body)\nboto.exception.EC2ResponseError: EC2ResponseError: 400 Bad Request\n<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Response><Errors><Error><Code>InvalidParameterValue</Code><Message>CIDR block /32 is malformed</Message></Error></Errors><RequestID>8f8f2151-5fc2-4a98-9b5a-82345f7c722d</RequestID></Response>\n", "module_stdout": "", "msg": "MODULE FAILURE"}
	to retry, use: --limit @/home/louisremi/Workspace/playbook/ansible-playbooks/openvpn.retry
```

**Comment by louisremi** [comment]
I'm still stuck with the following error:
```
TASK [ubuntu.raw_install_python : Update Apt.] *********************************
task path: /home/louisremi/Workspace/playbook/ansible-roles/ubuntu.raw_install_python/tasks/main.yml:3
<ec2-52-90-79-152.compute-1.amazonaws.com> ESTABLISH SSH CONNECTION FOR USER: ubuntu
<ec2-52-90-79-152.compute-1.amazonaws.com> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o 'IdentityFile="/home/louisremi/.ssh/chez-nestor"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ubuntu -o ConnectTimeout=10 -o ControlPath=/home/louisremi/.ansible/cp/ssh-%C -tt ec2-52-90-79-152.compute-1.amazonaws.com 'sudo -H -S -n -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-xwquasxjrkswlwjtaztmjkgqzuttafvy; which python || apt-get update'"'"''
fatal: [ec2-52-90-79-152.compute-1.amazonaws.com]: UNREACHABLE! => {
    "changed": false, 
    "msg": "Failed to connect to the host via ssh: ssh: connect to host ec2-52-90-79-152.compute-1.amazonaws.com port 22: Connection timed out\r\n", 
    "unreachable": true
}
```

**Comment by bonovoxly** [comment]
if you ran the playbook WITHOUT `public_ip=true`, the security group will NOT allow you to SSH in.  check out the first two tasks here:

https://github.com/bonovoxly/playbook/blob/master/ansible-roles/aws.securitygroups/tasks/main.yml

What this does is grab your public IP address and append it to the security group.  In fact, that's probably why you hit a security group bug previously. I'll correct that role.

So in summary, because you did not set `public_ip=true`, you cannot SSH into that system (the way the role is written).

**Comment by louisremi** [comment]
Ok. As soon as I add `public_ip=true`, the first error I encountered comes back:
```
TASK [aws.securitygroups : Create security groups.] ****************************
task path: /home/louisremi/Workspace/playbook/ansible-roles/aws.securitygroups/tasks/main.yml:25
Using module file /home/louisremi/Workspace/containers/env/local/lib/python2.7/site-packages/ansible/modules/core/cloud/amazon/ec2_group.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: louisremi
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo ~/.ansible/tmp/ansible-tmp-1487184536.22-112718468721145 `" && echo ansible-tmp-1487184536.22-112718468721145="` echo ~/.ansible/tmp/ansible-tmp-1487184536.22-112718468721145 `" ) && sleep 0'
<127.0.0.1> PUT /tmp/tmp6eUCej TO /home/louisremi/.ansible/tmp/ansible-tmp-1487184536.22-112718468721145/ec2_group.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/louisremi/.ansible/tmp/ansible-tmp-1487184536.22-112718468721145/ /home/louisremi/.ansible/tmp/ansible-tmp-1487184536.22-112718468721145/ec2_group.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/env python /home/louisremi/.ansible/tmp/ansible-tmp-1487184536.22-112718468721145/ec2_group.py; rm -rf "/home/louisremi/.ansible/tmp/ansible-tmp-1487184536.22-112718468721145/" > /dev/null 2>&1 && sleep 0'
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_mEQHJP/ansible_module_ec2_group.py", line 478, in <module>
    main()
  File "/tmp/ansible_mEQHJP/ansible_module_ec2_group.py", line 373, in main
    group.authorize(rule['proto'], rule['from_port'], rule['to_port'], thisip, grantGroup)
  File "/home/louisremi/Workspace/containers/env/local/lib/python2.7/site-packages/boto/ec2/securitygroup.py", line 203, in authorize
    dry_run=dry_run)
  File "/home/louisremi/Workspace/containers/env/local/lib/python2.7/site-packages/boto/ec2/connection.py", line 3207, in authorize_security_group
    params, verb='POST')
  File "/home/louisremi/Workspace/containers/env/local/lib/python2.7/site-packages/boto/connection.py", line 1227, in get_status
    raise self.ResponseError(response.status, response.reason, body)
boto.exception.EC2ResponseError: EC2ResponseError: 400 Bad Request
<?xml version="1.0" encoding="UTF-8"?>
<Response><Errors><Error><Code>InvalidParameterValue</Code><Message>Invalid integer value 0.0.0.0/0</Message></Error></Errors><RequestID>e3a19dab-5ff0-43e8-81b9-1a9810005ca8</RequestID></Response>

failed: [127.0.0.1] (item={u'rules': [{u'to_port': u'0.0.0.0/0', u'from_port': 1194, u'proto': u'udp'}], u'description': u'openvpn inbound rules', u'name': u'openvpn_inbound', u'tags': {u'Environment': u'dev', u'Name': u'openvpn_inbound'}}) => {
    "failed": true, 
    "invocation": {
        "module_name": "ec2_group"
    }, 
    "item": {
        "description": "openvpn inbound rules", 
        "name": "openvpn_inbound", 
        "rules": [
            {
                "from_port": 1194, 
                "proto": "udp", 
                "to_port": "0.0.0.0/0"
            }
        ], 
        "tags": {
            "Environment": "dev", 
            "Name": "openvpn_inbound"
        }
    }, 
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_mEQHJP/ansible_module_ec2_group.py\", line 478, in <module>\n    main()\n  File \"/tmp/ansible_mEQHJP/ansible_module_ec2_group.py\", line 373, in main\n    group.authorize(rule['proto'], rule['from_port'], rule['to_port'], thisip, grantGroup)\n  File \"/home/louisremi/Workspace/containers/env/local/lib/python2.7/site-packages/boto/ec2/securitygroup.py\", line 203, in authorize\n    dry_run=dry_run)\n  File \"/home/louisremi/Workspace/containers/env/local/lib/python2.7/site-packages/boto/ec2/connection.py\", line 3207, in authorize_security_group\n    params, verb='POST')\n  File \"/home/louisremi/Workspace/containers/env/local/lib/python2.7/site-packages/boto/connection.py\", line 1227, in get_status\n    raise self.ResponseError(response.status, response.reason, body)\nboto.exception.EC2ResponseError: EC2ResponseError: 400 Bad Request\n<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Response><Errors><Error><Code>InvalidParameterValue</Code><Message>Invalid integer value 0.0.0.0/0</Message></Error></Errors><RequestID>e3a19dab-5ff0-43e8-81b9-1a9810005ca8</RequestID></Response>\n", 
    "module_stdout": "", 
    "msg": "MODULE FAILURE"
}
```
:-(

**Comment by bonovoxly** [comment]
alright, i have a nasty little bug that i need to fix.  looks like the union of the empty rulesets is breaking my task.  didn't test enough.

if you want a quick fix, doing changing this line:
```
    rules: "{{ item.rules | union(bastionhost_ssh_rules) | union(current_public_ip_rules) }}"
```

to 

```
    rules: "{{ item.rules | union(current_public_ip_rules) }}"
```
might do the trick, until i get a fix.

**Comment by bonovoxly** [comment]
what version of ansible are you using?

did you happen to edit `vars/openvpn.yml`? it almost seems like you deleted some of the configuration.  when i deliberately break `vars/openvpn.yml`, by setting it to:

```
securitygroups:
  - name: openvpn_inbound
    description: 'openvpn inbound rules'
    rules:
      - proto: udp
        from_port: 1194
        to_port: 0.0.0.0/0
    tags:
      Name: "openvpn_inbound"
      Environment: "{{ env }}"
```

which is broken, the `to_port` is obviously not a CIDR, i can recreate the error you got.  Can you verify that variable?

**Comment by bonovoxly** [comment]
@louisremi I fixed up the documentation. let me know if it's helpful now.

**Comment by bonovoxly** [comment]
Going to close this.  If there are more issues, reopen and I will address them.

**Comment by louisremi** [comment]
Thanks a lot Bill, I'll look into this :-)
